### PR TITLE
Convert configuration to a class

### PR DESCRIPTION
**Pull request by MrCreosote**
This will permit adding config change listeners in a future commit.

Not considering whitespace changes will make this much easier to review.

### CODE REVIEW

File: src/utils/config.py [hunk 1]
```diff
@@ -2,87 +2,117 @@
...
+    if not _CONFIG_SINGLETON:
+        # could do this on module load, but let's delay until actually requested
+        _CONFIG_SINGLETON = Config()
+    _CONFIG_SINGLETON.reload(force_reload=force_reload)
+    return _CONFIG_SINGLETON
+
+
+class Config:
+    """
+    A class contining the configuration for the for the search indexer. Supports dict like
```
Comment (jayrbolton) [inline src/utils/config.py:29]: contining->containing

File: src/utils/config.py [hunk 2]
```diff
@@ -2,87 +2,117 @@
...
+                'admin_events': os.environ.get('KAFKA_ADMIN_TOPIC', 'indexeradminevents')
+            },
+            'config_timeout': 600,  # 10 minutes in seconds.
+            'last_config_reload': time.time(),
+            'proc_ready_path': proc_ready_path,  # File indicating the daemon is booted and ready
+            'generic_shard_count': os.environ.get('GENERIC_SHARD_COUNT', 2),
+            'generic_replica_count': os.environ.get('GENERIC_REPLICA_COUNT', 1),
+        }
+
+    def __getitem__(self, key):
```
Comment (jayrbolton) [inline src/utils/config.py:114]: nice

File: src/utils/config.py (hunk 1 above)
Comment (MrCreosote) [reply src/utils/config.py]: fixed

**Comment by MrCreosote** [comment]
> What about the logic of autoreloading the config on an interval?

> You could start a background thread on init. That way you could get rid of the expiration logic.

IMO it's easier to understand / reason about without the thread. I always try to avoid concurrency when I can, and threads open up the opportunity for weird concurrency bugs.

Also here https://github.com/kbase/index_runner_deluxe/issues/123#issuecomment-650427408
... I propose getting rid of the autoreloads altogether.

**Comment by jayrbolton** [comment]
To me, this:
> We continue checking GITHUB_RELEASE_URL 1 / minute for release changes and update on release change.
means autoreloading. What's the difference there?

**Comment by MrCreosote** [comment]
What I was referring to was the 10m reload. It would certainly be possible to have a thread to do the github release based reload, but my preference would be to leave it as is for the reasons already give.

**Comment by jayrbolton** [comment]
I thought that you had wanted to remove the config related code out of the event loop and into the config module, which in my mind included the release re-fetching logic. 

A non-thread way would be to add the logic (expiration check and reload) in the __getitem__ method

**Comment by MrCreosote** [comment]
> I thought that you had wanted to remove the config related code out of the event loop and into the config module, which in my mind included the release re-fetching logic.

I do want to do that. I just don't want to use a thread.

> A non-thread way would be to add the logic (expiration check and reload) in the getitem method

I was thinking of doing it in the `reload` method so it would happen when `config()` gets called, just like the prior way it was updated, but `getitem` works too. The former would match the previous update check trigger and be trivially faster, since you're checking for updates less often.

**Comment by jayrbolton** [comment]
I see, for some reason I was assuming that clients would use the Config class directly, but continuing to use the config function makes sounds good.

**Comment by jayrbolton** [comment]
Ok, LGTM

**Pull request merged**